plotly/tools.py

```diff
@@ -1443,6 +1443,376 @@ class FigureFactory(object):
+                                mode = "markers",
+                                name = name,
+                                marker = dict(
+                                    symbol = symbol,
+                                    size = size,
+                                    color = colors[color_index]),
+                                showlegend = False
+                                )
+
+                        # Push the trace into dictionary
+                        unique_leg_names[name] = trace
+                        if color_index >= (len(colors) - 1):
+                            color_index = -1
+                        color_index += 1
+                        
+                    trace_list.append(unique_leg_names)
+                    legend_param += 1
+
+            # Create list of index values for the axes
+            # eg. if dim = 3, then the indicies are [1, 2, 3]
+            indicies = range(dim)
+            indicies.remove(0)
+            indicies.append(dim)
+            j = 0
+            
+            for y_index in indicies:
+                for x_index in indicies:
+                    for name in trace_list[j]:
+                        fig.append_trace(trace_list[j][name], y_index, x_index)
+                    j += 1
+        
+            # Check if length of col_names is equal to the
+            # number of lists in matrix
+            if len(col_names) != dim:
+                raise exceptions.PlotlyError("Your list of variable_"
+                                             "names must match the "
+                                             "number of lists in your "
+                                             "array. That is to say that "
+                                             "both lists must have the "
+                                             "same dimension.")
+
+            # Insert col_names into the figure
+            for j in range(dim):
+                xaxis_place = "xaxis" + str(dim*dim - dim + 1 + j)
+                fig['layout'][xaxis_place].update(title = col_names[j])
+
+            for j in range(dim):
+                yaxis_place = "yaxis" + str(1 + dim*j)
+                fig['layout'][yaxis_place].update(title = col_names[j])
+
+            # Set height and width if not already selected by user
+            if (height == " "):
+                height = 400 + 200*(dim - 1)
+
+            if (width == " "):
+                width = 400 + 200*(dim - 1)
+
+            if diagonal == "Histogram":
+                fig['layout'].update(
+                    height = height, width = width,
+                    title = title,
+                    showlegend = True,
+                    barmode = "stack")
+                return fig
+
+            if diagonal == "Scatter":
+                fig['layout'].update(
+                    height = height, width = width,
+                    title = title,
+                    showlegend = True)
+                return fig
```

Comment:
If for some reason `diagonal` isn't `Histogram` or `Scatter`, `None` gets returned. This could be more robust using an `if`-`else` here.